source/common/compressor/zlib_compressor_impl.cc

```diff
@@ -31,32 +31,38 @@ void ZlibCompressorImpl::init(CompressionLevel comp_level, CompressionStrategy c
   initialized_ = true;
 }
 
-void ZlibCompressorImpl::flush(Buffer::Instance& output_buffer) {
-  process(output_buffer, Z_SYNC_FLUSH);
-}
-
 uint64_t ZlibCompressorImpl::checksum() { return zstream_ptr_->adler; }
 
-void ZlibCompressorImpl::compress(const Buffer::Instance& input_buffer,
-                                  Buffer::Instance& output_buffer) {
-  const uint64_t num_slices = input_buffer.getRawSlices(nullptr, 0);
+void ZlibCompressorImpl::compress(Buffer::Instance& buffer, State state) {
+  const uint64_t num_slices = buffer.getRawSlices(nullptr, 0);
   Buffer::RawSlice slices[num_slices];
-  input_buffer.getRawSlices(slices, num_slices);
+  buffer.getRawSlices(slices, num_slices);
 
   for (const Buffer::RawSlice& input_slice : slices) {
     zstream_ptr_->avail_in = input_slice.len_;
     zstream_ptr_->next_in = static_cast<Bytef*>(input_slice.mem_);
-    process(output_buffer, Z_NO_FLUSH);
+    process(buffer, Z_NO_FLUSH);
```

Comment:
Can you add an ASSERT into `process` that confirms that in the case of Z_NO_FLUSH no data is added to the buffer? Is that always guaranteed? I'm still a little concerned about the in/out buffer behavior here. Otherwise can you add more comments?

Reply:
It turns out if with RLE, SPEED and 1 as memory level, I can see data is being added to the buffer when flush_state == Z_NO_FLUSH. Will try to debug it.

Reply:
Hi @mattklein123 You probably saw something that we didn't see. Question, even if data is added to the buffer, what the consequences would be?

Reply:
I think it will work, because you add data to the end, and you drain from the beginning. Also, you snap the slices before you start adding, so you should never drain compressed data. However, it's not super intuitive how this works, so if you keep it this way I would add a bunch of comments (I think it's probably fine w/ comments).

Reply:
I agree that this is not the most intuitive design. I'm planning on revisiting this implementation as soon as I get some stuff out of my way. What I and @dio agreed, was to go for correctness now. We will add the comments though. Thanks for the input.